docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+6. This has no consequences on the config size because `a` and `c` are the only services in the mesh
+7. Now let's bring service `d` into the mesh
+8. When we do that we will observe the first consequence of config trimming,
+   service `d` will have only `a` cluster but won't have `c` cluster because we defined that `c` can only be contacted by `a`
+9. We repeat this until we have all services in the mesh, all new services that join already have a smaller config because we trimmed it along the way
+
+It's critical to remember to only switch to `deny` when **ALL** consumers are **defined in MTP** and are in the Mesh.
+
+There are other concerns that also need addressing:
+- breaking "top-level targetRef selects proxy to configure" rule
+  - I don't think we can do anything about this, it's at the core of this feature working, what we can (and should) do
+    is to have this feature behind a flag and if there is too much negative feedback about this point we can revert it
+- wrong errors on the client side, instead of 403 client will get 404/"no upstream"
+  - in public facing APIs this is quote common to get a 404 to a resource that you don't have permissions to
+  - I imagine we could have a flag to distinguish between a regular 404 and this 404 caused by no upstream resulting from config trimming and return a 403, but
+    I think this is out of scope for this, probably a future improvement
+- wrong RBAC stats on the server side
+  - we would have correct stats on the client side - if we can translate that and surface it to the user I don't think anyone would mind, but
+    I think its is for future improvement
+- assuming that every service owner knows which services consume their API
+  - as a future improvement we will come up with tooling to assist the users with this problem
+    - have metrics / logs and a dashboard when `allow_with_shadow_deny` is used (this is not a silver bullet - it assumes that during "discovery" period all possible communication will happen, if you gather data for a week and there is a job that runs every quarter it probably won't be recorded)
+
+### Dynamically load clusters using ODCDS
+
+ODCDS only works with `host` header, so it would only work for HTTP traffic.
+
+Based on that limitation we could have in our DNS component a special IP for services with `kuma.io/protocol: http`
+
+```yaml
+inlineDnsTable:
+  virtualDomains:
+  - answerTtl: 30s
+    endpoint:
+      addressList:
+        address:
+        - 241.0.0.0 # normally it's 240.0.0.0
+    name: some-service.svc
+```
+
+And that I would have a special listener in Envoy that would have an HTTP filter that would use `odcds`.
+
+Pros/cons:
+* This is a very clean solution that would require minimal effort from the users, just marking services with `kuma.io/protocol: http`
+* It requires implementing `odcds` in `go-control-plane` and changes in Kuma
+* No solution for TCP traffic
+* Not possible in the required timeframe
```

Comment:
I think a feature like this is by far our best hope. The advantage of MTP is that we would be able to fail ODCDS if we request a cluster we don't have access too so if we implement this later we'd still need the MTP feature.

Reply:
> The advantage of MTP is that we would be able to fail ODCDS if we request a cluster we don't have access too so if we implement this later we'd still need the MTP feature.

I don't think I understand this fully, so you'd want to `nack` the ODCDS request if someone requests a cluster they don't have access to?

Reply:
Yeah I'd guess so no? Or have a cluster with a direct response? Anyway out of scope here :)

Reply:
Yeah, cluster with direct response would work I think.

Reply:
I would add cons: delay on the first request. It seems like for some clients it might be a blocker. 
Also, another thing, what if your CP is down and you want to make a request to a new service? In this case, MTP has the knowledge about services already while ODCDS won't get it?

Reply:
> ODCDS won't get it?

yeah you won't get a response so it won't work

added the cons mentioned